### PR TITLE
Query information should be added to the return

### LINKED ISSUE #96: Add query info to return
**Is your feature request related to a problem? Please describe.**
Query information should be added to the return to provide more transparency

See #90 for related notes.

### PR DESCRIPTION
**Pull request by vchendrix**
Adds query information to the return to provide more
transparency
  + Adds new schema package for parsing and validating query inputs
    (See basin3d.core.schema.query)
  + Refactors basin3d.core.types used in queries as Enum objects to be
    used in pydantic query schemas (See basin3d.core.schema.enum)
  + Updates requirements.txt with pydantic dependency
  + Refactors classes in basin3d.core.synthesis
     - to use new pydantic schema data classes
     - return new SynthesisResponse object containing information about
       the original query.
  + Refactors tests for the described changes

Closes #96

## Type of change
Although this is a breaking change, the api call that outputs HDF files should not be affected

- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- [X] Unit tests
- [X] Integration tests 
- [X] Test coverage >= 90%
- [X] Flake8 Tests 
- [X] Mypy Tests 

### Test Configuration
* Python Version: 3.8

## PR Self Evaluation
strikethrough things that don’t make sense for your PR

- [X] My code follows the agreed upon best practices
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests or modified existing tests that prove my fix is effective or that my feature works
- [X] Existing unit tests pass locally with my changes
- ~Any dependent changes have been merged and published in the appropriate modules~
- [X] I have performed a self-review of my own code

### CODE REVIEW

File: basin3d/plugins/usgs.py [hunk 1]
```diff
@@ -420,36 +404,39 @@ def get_hydrological_unit_codes(self):
...
+            mf_query = QueryMonitoringFeature(monitoring_features=[query.id], feature_type=FeatureTypeEnum.POINT)
+
+        for o in self.list(query=mf_query):
             return o
+
+        # An 8 character code can also be a point, Try that
+        if len(query.id) == 8:
+            for o in self.list(query=QueryMonitoringFeature(monitoring_features=[query.id],
+                                                            feature_type=FeatureTypeEnum.POINT)):
+                return o
```
Comment (dschristianson) [inline basin3d/plugins/usgs.py:439]: OK for now. I'd like to do a bit more testing. I thought we had an issue for this but am not seeing it. I'd like to make one and reference it in this PR or commit history (if not already).

File: basin3d/synthesis.py [hunk 2]
```diff
@@ -337,74 +339,57 @@ def monitoring_features(self, id: str = None, feature_type: str = None, datasour
         USGS-13010450 [(-110.5874305, 43.9038296)]
         ...
 
-        :param id: Unique feature identifier
-        :param feature_type: feature type
-        :param datasource: Datasource id prefix (e.g USGS)
-        :param monitoring_features: List of monitoring feature identifiers (eg. USGS-0010)
-        :param parent_features: List of parent monitoring features to search by
+        :param query:
```
Comment (dschristianson) [inline basin3d/synthesis.py:342]: Add query definition. From other docsrings: "The query information for this function"

File: basin3d/synthesis.py [hunk 3]
```diff
@@ -659,7 +636,8 @@ def get_timeseries_data(synthesizer: DataSynthesizer, location_lat_long: bool =
             shutil.rmtree(working_directory)
 
 
-def _output_df(output_directory, output_name, query_info, metadata_store, first_timestamp, last_timestamp, temporal_resolution) -> \
+def _output_df(output_directory, output_name, query_info, metadata_store, first_timestamp, last_timestamp,
+               temporal_resolution) -> \
```
Comment (dschristianson) [inline basin3d/synthesis.py:640]: minor: weird line break??

File: basin3d/synthesis.py [hunk 4]
```diff
@@ -706,7 +684,8 @@ def _output_df(output_directory, output_name, query_info, metadata_store, first_
                                 query=query_info)
 
 
-def _output_hdf(output_directory: str, output_name: str, query_info: QueryInfo, metadata_store: dict, first_timestamp, last_timestamp,
+def _output_hdf(output_directory: str, output_name: str, query_info: QueryInfo, metadata_store: dict, first_timestamp,
+                last_timestamp,
```
Comment (dschristianson) [inline basin3d/synthesis.py:688]: minor: weird line break?

**Review by dschristianson — COMMENTED** [review]
Minor stuff -- see what you want to do with it.

File: basin3d/core/schema/enum.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    PARTIALLY_CHECKED = "PARTIALLY_CHECKED"
+
+
+class StatisticEnum(str, BaseEnum):
+    """Enumeration for Statistics"""
+    INSTANT = "INSTANT"
+    MEAN = "MEAN"
+    MIN = "MIN"
+    MAX = "MAX"
+    TOTAL = "TOTAL"
```
[the remark this answers is not in the document]
Comment (vchendrix) [reply basin3d/core/schema/enum.py:105]: I think I turned this off in our flake8 checking. Are you seeing this when running tests?

File: basin3d/synthesis.py (hunk 2 above)
Comment (vchendrix) [reply basin3d/synthesis.py:342]: Good point.  I thought about that but forgot to update the docs.

File: basin3d/synthesis.py (hunk 3 above)
Comment (vchendrix) [reply basin3d/synthesis.py:640]: Just formatting. I can remove but it is harmless.

File: basin3d/core/synthesis.py [hunk 6]
```diff
@@ -211,49 +234,30 @@ class MonitoringFeatureAccess(DataSourceModelAccess):
...
+        if query:
+            id_prefix = plugin_access.datasource.id_prefix
+            if query.monitoring_features:
+                synthesized_query.monitoring_features = _synthesize_query_identifiers(
+                    values=query.monitoring_features,
+                    id_prefix=id_prefix)
+            if query.parent_features:
+                synthesized_query.parent_features = _synthesize_query_identifiers(
+                    values=query.parent_features,
+                    id_prefix=id_prefix)
```
[the remark this answers is not in the document]
Comment (vchendrix) [reply basin3d/core/synthesis.py:258]: I am not sure I understand why we would need to do this.

File: basin3d/plugins/usgs.py (hunk 1 above)
Comment (vchendrix) [reply basin3d/plugins/usgs.py:439]: When we had our review, I had forgotten that I had already done this.  I don't remember if an issue was created...

File: basin3d/core/synthesis.py [hunk 7]
```diff
@@ -305,28 +309,26 @@ def synthesize_query_params(self, plugin_access: DataSourcePluginAccess, **kwarg
...
-            observed_property_variables = query_params.get(QUERY_PARAM_OBSERVED_PROPERTY_VARIABLES, [])
-            query_params[QUERY_PARAM_OBSERVED_PROPERTY_VARIABLES] = [o.datasource_variable for o in
-                                                                     plugin_access.get_observed_properties(
-                                                                         observed_property_variables)]
+        synthesized_query = query.copy()
+
+        if query:
+
+            if query.monitoring_features:
+                synthesized_query.monitoring_features = _synthesize_query_identifiers(values=query.monitoring_features,
```
[the remark this answers is not in the document]
Comment (vchendrix) [reply basin3d/core/synthesis.py:323]: same here as above. I am not sure why we need to do this.

File: basin3d/core/schema/enum.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/schema/enum.py:105]: fine to ignore

File: basin3d/core/schema/query.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        alias_generator = _to_camelcase
+        # whether an aliased field may be populated by its name as given by the model attribute
+        #  (allows bot camelcase and underscore fields)
+        allow_population_by_field_name = True
+        # Instead of using enum class use enum value (string object)
+        use_enum_values = True
+        # Validate all fields when initialized
+        validate_all = True
+        # Allows generic object to be used for data field
+        arbitrary_types_allowed = True
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/schema/query.py:148]: fine to ignore

File: basin3d/synthesis.py (hunk 3 above)
Comment (dschristianson) [reply basin3d/synthesis.py:640]: just jarring to read. I had to look twice to make sure there wasn't something amiss. As mentioned it is minor.

File: basin3d/core/synthesis.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/synthesis.py:258]: Helpful to future developers. It took me a bit to remember why we needed to do this for monitoring features and parent features and not feature type.

File: basin3d/synthesis.py (hunk 3 above)
Comment (vchendrix) [reply basin3d/synthesis.py:646]: k. Yeah. this is a function of the autoformatting

File: basin3d/synthesis.py (hunk 4 above)
Comment (vchendrix) [reply basin3d/synthesis.py:694]: This is actually formatting to keep the line length reasonable.

File: basin3d/synthesis.py (hunk 3 above)
Comment (vchendrix) [reply basin3d/synthesis.py:646]: same as below. it is geared to keep the line length reasonable. It is common practice but of course subject to opinion! :)

**Pull request merged**